File: src/main/java/io/eontimer/service/factory/timer/DelayTimerFactory.java
package io.eontimer.service.factory.timer;

import java.util.Arrays;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import io.eontimer.model.timer.TimerConstants;
import io.eontimer.service.CalibrationService;
import io.eontimer.util.TimeUtil;

@Service
public class DelayTimerFactory {

	@Autowired
	private SecondTimerFactory secondTimer;

	@Autowired
	private CalibrationService calibrationService;

	public List<Long> createStages(long targetSecond, long targetDelay, long calibration) {
		return Arrays.asList(stage1(targetSecond, targetDelay, calibration), stage2(targetDelay, calibration));
	}

	public long calibrate(long targetDelay, long delayHit) {
		final long delta = calibrationService.toMillis(delayHit) - calibrationService.toMillis(targetDelay);
		if (Math.abs(delta) <= TimerConstants.CLOSE_THRESHOLD) {
			return (long) (TimerConstants.CLOSE_UPDATE_FACTOR * delta);
		}
		return ((long) TimerConstants.UPDATE_FACTOR) * delta;
	}

	private long stage1(long targetSecond, long targetDelay, long calibration) {
		return TimeUtil.toMinimumLength(secondTimer.createStages(targetSecond, calibration).get(0) - calibrationService.toMillis(targetDelay));
	}

	private long stage2(long targetDelay, long calibration) {
		return calibrationService.toMillis(targetDelay) - calibration;
	}

}

File: src/service/factory/DelayTimerFactory.js
import secondTimerFactory from "./SecondTimerFactory";
import calibrationUtil from "../util/CalibrationUtil";
import timeUtil from "../util/TimeUtil";

const createStages = (targetSecond = 0, targetDelay = 0, calibration = 0, handheld) => {
  return [stage1(targetSecond, targetDelay, calibration, handheld), stage2(targetDelay, calibration, handheld)];
};

const calibrate = (targetDelay = 0, delayHit = 0, handheld) => {
  const delta = calibrationUtil.toMillis(delayHit, handheld) - calibrationUtil.toMillis(targetDelay, handheld);
  if (Math.abs(delta) <= 167) {
    return 0.75 * delta;
  }
  return 1.0 * delta;
};

const stage1 = (targetSecond, targetDelay, calibration, handheld) => {
  return timeUtil.toMinimumLength(secondTimerFactory.createStages(targetSecond, calibration)[0] - calibrationUtil.toMillis(targetDelay, handheld));
};

const stage2 = (targetDelay, calibration, handheld) => {
  return calibrationUtil.toMillis(targetDelay, handheld) - calibration;
};

export default {
  createStages,
  calibrate,
};
